2002 Segment contours

plt-joey opened the pull request:
- Adds features described in #2002

plt-joey committed:
Initial contour worker

plt-joey committed:
Add segment contour rendering.

ivmartel commented:
Looking good! I'm just missing a similar behaviour in all planes: when you switch to MPR and draw, the segmentation looses its transparency.

plt-joey committed:
Fix contours on MPR view

plt-joey commented:
> Looking good! I'm just missing a similar behaviour in all planes: when you switch to MPR and draw, the segmentation looses its transparency.

Woops, forgot about orthogonal views. Fixed now.
Regarding your comment on the contours looking strange on odd pixel scales, I agree, the problem is there isn't really a nice way to fix that for thin contours without completely changing how the contours are drawn and calculated.

plt-joey committed:
Bound contour distance to avoid overflow

plt-joey committed:
Fix lint

ivmartel commented:
Looking better on orthogonal views thanks. This will be enough for now, it looks good for data with good resolution. 

Regarding the contour thickness, the code says that it needs to be >= 1, would it be complicated to support a 0 thickness for people who do not want a border?

plt-joey committed:
Allow 0 size contour on demo page, fix doc comment

plt-joey commented:
> Looking better on orthogonal views thanks. This will be enough for now, it looks good for data with good resolution.
> 
> Regarding the contour thickness, the code says that it needs to be >= 1, would it be complicated to support a 0 thickness for people who do not want a border?

That documentation comment was incorrect, it should have said `>= 0`, 0 thickness is already supported. I updated the documentation comments and also updated the slider on the demo page to go down to 0.

To get the same rendering behaviour that segments used to have; set the contour thickness to 0 and the fill opacity to 100%. That will remove the border and make the segmentation opacity controlled by the global opacity of the layer.

ivmartel commented:
The interaction is ok on most of the data I tried (for example the bbmri in the tests/data folder) meaning you can paint+move (at a reasonable speed) and the segments follow your moves. But I'm having troubles with MR data (see attached [IM-0001-0001.zip](https://github.com/user-attachments/files/22851216/IM-0001-0001.zip) from some old Osirix demo data). The first paint takes a bit of time, paint+move has troubles following. Not sure what is the problem, could you look into it?

ivmartel commented:
Another data where paint quite slow: [nema-mr1_jply.dcm](https://github.com/ivmartel/dwv/blob/develop/tests/data/nema-mr1_jply.dcm) (from the tests/data folder)

plt-joey committed:
Contrain contours to a max size for performance

plt-joey commented:
> The interaction is ok on most of the data I tried (for example the bbmri in the tests/data folder) meaning you can paint+move (at a reasonable speed) and the segments follow your moves. But I'm having troubles with MR data (see attached [IM-0001-0001.zip](https://github.com/user-attachments/files/22851216/IM-0001-0001.zip) from some old Osirix demo data). The first paint takes a bit of time, paint+move has troubles following. Not sure what is the problem, could you look into it?

> Another data where paint quite slow: [nema-mr1_jply.dcm](https://github.com/ivmartel/dwv/blob/develop/tests/data/nema-mr1_jply.dcm) (from the tests/data folder)

Ah, the default brush size on those scan is MUCH larger than what I was testing with. I didn't realize it was based on the spacing and not the pixels.

I fixed it, but I had to constrain the max contour size to 10px (set as a constant in `view.js`) to get the brush that big to be performant. We could maybe get a little bit faster with a better algorithm, but anything much faster than that would either introduce visual artifacts (eg, async batching), or would have to run on the GPU (or some other native API).

ivmartel commented:
Ok, speed seems good enough, thanks. Do you think you could add some tests?

plt-joey committed:
Add contour unit tests

plt-joey commented:
> Ok, speed seems good enough, thanks. Do you think you could add some tests?

Added and rebased

ivmartel committed:
Merge branch 'develop' into segment-contours

ivmartel committed:
Remove trailing spaces

ivmartel commented:
Thanks!
